sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -1912,6 +1913,26 @@ class Analyzer(
           nondeterToAttr.get(e).map(_.toAttribute).getOrElse(e)
         }.copy(child = newChild)
 
+      case j: Join if j.condition.isDefined && !j.condition.get.deterministic =>
+        j match {
+          // We can push down non-deterministic joining keys.
+          // We can't push down non-deterministic conditions.
+          case ExtractEquiJoinKeys(_, leftKeys, rightKeys, conditions, _, _)
```

Comment:
I do not think we can have an easy solution to ensure it always works as you expected. `EnsureRequirements` is just one of rules that could break it.

The other preceding join conditions before equi join condition also could impact it. It could be skipped if the preceding join conditions is false, right?